File: Übungsblätter/Blatt 6/Einreichungen/Tao Xiang/DocumentCollection.java

public class DocumentCollection {
	DocumentCollectionCell anfang;
	int anzahl; // number of documents in this collection
	public DocumentCollection()
	{
		anfang=null;
		anzahl=0;
	}
	public void prependDocument(Document doc)
	{
		if(doc!=null) {
			anzahl++;
			if(anfang!=null) {
				DocumentCollectionCell tmp=this.anfang;
				for(int i=0;i<anzahl-1;i++)
				{
					tmp.setIndex(i+1);
					tmp=tmp.getNextDcCCell();
				}
		DocumentCollectionCell newbeginn=new DocumentCollectionCell(doc);
		newbeginn.setNextDcCCell(anfang);
		this.anfang.setLastDcCCEll(newbeginn);
		this.anfang=newbeginn;
		this.anfang.setIndex(0);
		
		
			}
			if(anfang==null)
			{
				anfang= new DocumentCollectionCell(doc);
				anfang.setIndex(0);
			}
		
		}
	}
	public void appendDocument(Document doc)
	{
		if(doc!=null)
		{
			anzahl++;
			if(anfang!=null) {
			DocumentCollectionCell newend=new DocumentCollectionCell(doc);
			DocumentCollectionCell temp=this.anfang;
			while(temp.getNextDcCCell()!=null)
			{
				temp=temp.getNextDcCCell();
			}
			temp.setNextDcCCell(newend);
			newend.setLastDcCCEll(temp);
			newend.setIndex(anzahl);
			}
			if(anfang==null) {
				anfang=new DocumentCollectionCell(doc);
				anfang.setIndex(0);
			}
		}
	}
	private DocumentCollectionCell getDocCC(int index)
	{
		return this.findTheDocCCAtIndex(index);
	}
	public boolean isEmpty()
	{
		return(anzahl==0);
	}
	public int numDocuments()
	{
		return anzahl;
	}
	public Document getFirstDocument()
	{
		if(this.anfang==null) return null;
		else {
			return this.anfang.getDoc();
		}
	}
	private DocumentCollectionCell findTheLastDocCC()
	{
		if(this.anfang!=null) {
		DocumentCollectionCell temp=this.anfang;
		while(temp.getNextDcCCell()!=null)
		{
			temp=temp.getNextDcCCell();
		}
		return temp;}
		else return null;
	}
	private DocumentCollectionCell findTheDocCCAtIndex(int index)
	{
		if(index>=0&&index<this.anzahl) {
			DocumentCollectionCell temp=this.anfang;
			for(int i=1;i<=index;i++)
			{
				temp=temp.getNextDcCCell();
			}
			return temp;}
			else return null;
	}
	public Document getLastDocument()
	{
		if(this.anfang!=null)
		return this.findTheLastDocCC().getDoc();
		else return null;
	}
	public void removeFirstDocument()
	{
		if(this.anzahl!=0)
		{
			anzahl--;
			DocumentCollectionCell tmp =this.anfang;
			this.anfang=this.anfang.getNextDcCCell();
			anfang.setLastDcCCEll(null);
			tmp.setNextDcCCell(null);
		}
		DocumentCollectionCell tmp2=this.anfang;
		for(int i=0;i<anzahl;i++)
		{
			tmp2.setIndex(i);
			tmp2=tmp2.getNextDcCCell();
		}
	}
	public void removeLastDocument()
	{
		if(this.anzahl!=0)
		{
			anzahl--;
 
			DocumentCollectionCell tmp;
			tmp=this.findTheLastDocCC().getLastDcCCell();
			this.findTheLastDocCC().setLastDcCCEll(null);
			tmp.setNextDcCCell(null);
			
			
		}
	}
	public boolean remove(int index)
	{
		if(index<0||index>=this.anzahl) return false;
		else {
			anzahl--;
			DocumentCollectionCell temp= this.findTheDocCCAtIndex(index);
			DocumentCollectionCell tmp2=this.findTheDocCCAtIndex(index+1);
			temp.getLastDcCCell().setNextDcCCell(temp.getNextDcCCell());
			temp.getNextDcCCell().setLastDcCCEll(temp.getLastDcCCell());
			temp.setLastDcCCEll(null);
			temp.setNextDcCCell(null);
			for(int i=0;i+index<anzahl;i++)
			{
				tmp2.setIndex(i+index);
				tmp2=tmp2.getNextDcCCell();
			}
			
			
			return true;
		}
	}
	public Document get(int index)
	{
		if(index<0||index>=this.anzahl) return null;
		else {
			return this.findTheDocCCAtIndex(index).getDoc();
		}
	}
	public int indexof(Document doc)
	{
		int ret=-1;
		if(this.anzahl!=0)
		{
			DocumentCollectionCell tmp=this.anfang;
			for(int i=0;i<anzahl;i++,tmp=tmp.getNextDcCCell())
			{
				if(tmp.getDoc().equals(doc))
				{
					ret=i;
					break;
				}
			}
		}
		return ret;
	}
	public boolean contains(Document doc)
	{
		boolean contains=false;
		if(this.anzahl!=0)
		{
			for(int index=0;index<anzahl;index++)
			{
				if(this.findTheDocCCAtIndex(index).getDoc().equals(doc))
				{
					contains=true;
					break;
				}
			}
		}
		return contains;
	}
	private  static WordCountsArray merge(WordCountsArray wc1,WordCountsArray wc2)
	{
		WordCountsArray ret =new WordCountsArray(wc1.size()+wc2.size());
		for(int i=0;i<wc1.size();i++)
		{
			ret.add(wc1.getWord(i),wc1.getCount(i) );
		}
		for(int i=0;i<wc2.size();i++)
		{
			ret.add(wc2.getWord(i),wc2.getCount(i) );
		}
		
		return ret; 
				
	}
	private WordCountsArray allWords()
	{
		if(this.anzahl==0) return null;
		else {
			WordCountsArray ret = new WordCountsArray(10);
			DocumentCollectionCell tmp =this.anfang;
			for(int i=0;i<this.anzahl;i++) {
			ret=merge(ret,tmp.getDoc().getWordCounts());
			tmp=tmp.getNextDcCCell();}
			return ret;
		}
	}
	private void addZeroWordsToDocuments()
	{
		WordCountsArray allwords=this.allWords();
		DocumentCollectionCell tmp= this.anfang;
		for(int j=0;j<this.anzahl;j++) {
			
		for(int i=0;i<allwords.size();i++)
		{
			int index=tmp.getDoc().getWordCounts().getIndexOfWord(allwords.getWord(i));
			if(index==-1)
			{
				tmp.getDoc().getWordCounts().add(allwords.getWord(i), 0);
			}
		}
		tmp=tmp.getNextDcCCell();
		}
	}
	public void match(String searchQuery)
	{
		this.addZeroWordsToDocuments();
		Document temp=new Document(searchQuery);
		this.prependDocument(temp);
		this.addZeroWordsToDocuments();
		DocumentCollectionCell tmp=this.anfang;
		for(int i=0;i<this.anzahl;i++)
		{
			tmp.getDoc().getWordCounts().sort();
			tmp=tmp.getNextDcCCell();
		}
		DocumentCollectionCell tmp2=this.anfang.getNextDcCCell();
		for(int i=1;i<this.anzahl;i++)
		{
			tmp2.setSimilarity(this.anfang.getDoc().getWordCounts().computeSimilarity(tmp2.getDoc().getWordCounts()));
			tmp2=tmp2.getNextDcCCell();
		}
		this.removeFirstDocument();
		this.sortBySimilarityDesc();
	}
	public double getQuerySimilarity(int index)
	{
		return this.findTheDocCCAtIndex(index).getSimilarity();
	}
	private static void merge(DocumentCollection docC,int l1,int r1,int l2, int r2)
	{
		DocumentCollectionCell temp1 = null,temp2=null;
		if(docC.findTheDocCCAtIndex(l1).getLastDcCCell()!=null)
		{
			temp1=docC.findTheDocCCAtIndex(l1).getLastDcCCell();
		}
		if(docC.findTheDocCCAtIndex(r2).getNextDcCCell()!=null)
		{
			temp2=docC.findTheDocCCAtIndex(r2).getNextDcCCell();
		}
		int p1=l1,p2=l2;
		DocumentCollectionCell[] tmp =new DocumentCollectionCell[docC.anzahl];
		int index=l1;
		int index2=l1-1;
		while(p1<=r1&&p2<=r2)
		{
		if(docC.findTheDocCCAtIndex(p1).getSimilarity()>=docC.findTheDocCCAtIndex(p2).getSimilarity())
			{
				tmp[index]=docC.findTheDocCCAtIndex(p1);
				tmp[index].setIndex(++index2);
				index++;
				p1++;
			}
			else {
				tmp[index]=docC.findTheDocCCAtIndex(p2);
				tmp[index].setIndex(++index2);
				index++;
				p2++;
			}
		}
		while(p1<=r1) {
			tmp[index]=docC.findTheDocCCAtIndex(p1++);
			tmp[index].setIndex(++index2);
			index++;
		}
		while(p2<=r2)
		{ 
		
			tmp[index]=docC.findTheDocCCAtIndex(p2++);
		    tmp[index].setIndex(++index2);
			index++;
		}
		
		for(int i=l1;i+1<=r2;i++)
		{
	
			tmp[i].setNextDcCCell(tmp[i+1]);
			tmp[i+1].setLastDcCCEll(tmp[i]);
		}
		
		if(temp1!=null) { temp1.setNextDcCCell(tmp[l1]);tmp[l1].setLastDcCCEll(temp1);}
	    if(temp2!=null) {tmp[r2].setNextDcCCell(temp2);temp2.setLastDcCCEll(tmp[r2]);}
		
	    for(int i=l1;i<=r2;i++)
		{
			if(tmp[i].getIndex()==0)
			{
				docC.anfang=tmp[i];
				docC.anfang.setLastDcCCEll(null);
				break;
			}
		}	
	}
	private static void mergeSort(DocumentCollection docC,int left,int right)
	{
		if(left<right)
		{
			int mid=(left+right)/2;
			mergeSort(docC,left,mid);
			mergeSort(docC,mid+1,right);
			merge(docC,left,mid,mid+1,right);
			
		}
	}
	private void sortBySimilarityDesc()
	{
		DocumentCollection tmp=this;
		mergeSort(tmp,0,tmp.anzahl-1);
	}
	
	
	public static void main(String[] args)
	{
		String anfrage="hallo ich bin Tao";
		Document a=new Document("hallo");
		Document b=new Document("hallo ich");
		Document c= new Document("hallo ich bin Tao");
		Document d=new Document("sdsadsadsadsadasddasd");
		Document h= new Document("ich liebe dich");
		Document g =new Document("ich dich");
		Document l=new Document("servus ich b");

		DocumentCollection i=new DocumentCollection();
		i.appendDocument(a);
		i.appendDocument(b);
		i.appendDocument(c);
		i.appendDocument(d);
		i.appendDocument(h);
		i.appendDocument(g);
		i.appendDocument(l);
		i.match(anfrage);
		for(int j=0;j<i.anzahl;j++)
		{
			System.out.println(i.getDocCC(j).getSimilarity());
		}
		//i.sortBySimilarityDesc();
		
		/*i.match(anfrage);
		/*for(int j=0;j<i.anzahl;j++)
		{
			System.out.println(i.getQuerySimilarity(j));
		}}*/
}}

File: search_machine/_documentcollection.py
from ._document import Document
from ._wordcount import WordCountList

class DocumentCollectionCell:
    def __init__(self, doc: Document):
        assert doc is not None
        self.__document__ = doc
        self.__next__ = None

    def getDocument(self) -> Document:
        return self.__document__
    
    def setNext(self,o: object):
        if not isinstance(o,self.__class__):
            raise RuntimeError(f'The given object is not from class {self.__class__}.')
        self.__next__ = o
    def getNext(self):
        return self.__next__

    def __str__(self) -> str:
        return str(self.__document__)


class DocumentCollection:
    def __init__(self) -> None:
        self.__start__ = None
        self.__end__ = self.__start__
        self.__num__ = 0
        self.similarities = None
        
    def __assert__(self):
        if self.__num__ > 0:
            if not (isinstance(self.__start__,DocumentCollectionCell) and isinstance(self.__end__,DocumentCollectionCell)):
                raise RuntimeError(f'The start cell and end cell is not a DocumentCollectionCell!')
    
    def prependDocument(self, doc: Document):
        if doc is None:
            return
        
        if self.__start__ is None:
            self.__start__ = DocumentCollectionCell(doc)
            self.__end__ = self.__start__
        
        else:
            newcell = DocumentCollectionCell(doc)
            newcell.setNext(self.__start__)
            self.__start__ = newcell

        self.__num__ += 1
        self.__assert__()
        
    
    def appendDocument(self,doc:Document):
        if doc is None:
            return
        
        if self.__start__ is None:
            self.__start__ = DocumentCollectionCell(doc)
            self.__end__ = self.__start__
        else:
            self.__end__.setNext(DocumentCollectionCell(doc=doc))
            self.__end__ = self.__end__.getNext()

        self.__num__ += 1
        self.__assert__()
            
    def isEmpty(self)-> bool:
        return self.__num__ == 0
    
    def __len__(self):
        return self.__num__
    
    def getFirstDocument(self):
        return self.__start__.getDocument()
    
    def getLastDocument(self):
        return self.__end__.getDocument()
    
    def removeFirstDocument(self):
        if self.__num__ == 0:
            return

        if self.__num__ > 1:
            self.__start__ = self.__start__.getNext()
        elif self.__num__ == 1:
            self.__start__ = None
            self.__end__ = None

        self.__num__ -= 1
        self.__assert__()

    def getDocumentCellAt(self,idx:int) -> DocumentCollectionCell:
        if idx < 0 or idx >= self.__num__:
            raise IndexError('Index out of range.')
        else:
            tmp = self.__start__
            for idx in range(idx):
                tmp = tmp.getNext()
            return tmp


    def removeLastDocument(self):
        if self.__num__ == 0:
            return
        if self.__num__ > 1:
            tmp = self.getDocumentCellAt(self.__num__-2)
            self.__end__ = tmp
            tmp.__next__= None

        if self.__num__ == 1:
            self.__start__ = None
            self.__end__ = None

        self.__num__ -= 1
        self.__assert__()

    def remove(self,idx:int) -> bool:
        if idx < 0 or idx >= self.__num__:
            return False
        elif idx == 0:
            self.removeFirstDocument()
        elif idx == self.__num__-1:
            self.removeLastDocument()
        else:
            cell = self.getDocumentCellAt(idx-1)
            cell.setNext(cell.getNext().getNext())
            self.__num__ -= 1

        self.__assert__()

    def get(self,idx: int):
        if idx >= 0:
            return self.getDocumentCellAt(idx).getDocument()
        else:
            idx += self.__len__()
            return self.getDocumentCellAt(idx).getDocument()

    def __iter__(self):
        self.__pointer__ = self.__start__
        return self

    def __next__(self):
        if self.__pointer__ is None:
            raise StopIteration
        else:
            tmp = self.__pointer__
            self.__pointer__ = self.__pointer__.getNext()
            return tmp


    def indexOf(self, doc: Document):
        for id,cell in enumerate(self):
            if cell.getDocument() == doc:
                return id
        return -1

    def __contains__(self, item):
        if not isinstance(item,Document):
            return False
        else:
            for cell in self:
                if cell.getDocument() == item:
                    return True

            return False

    def allWords(self) -> WordCountList:
        ret = WordCountList()
        for cell in self:
            doc = cell.getDocument()
            wordcounts = doc.getWordCounts() # wordcountslist
            ret.extend(wordcounts)
        return ret


    def addZeroWordsToDocuments(self):
        allwords = self.allWords()
        for wc in allwords:
            word = wc.word
            for cell in self:
                cell.getDocument().addZeroCountWord(word)

    def match(self, searchQuery: str):
        query_doc = Document(None,searchQuery,None,None,None,None)
        self.appendDocument(query_doc)
        self.addZeroWordsToDocuments()
        similarities = []
        for i in range(len(self)-1):
            cell = self.getDocumentCellAt(i)
            similarities.append(cell.getDocument().getWordCounts().computeSimilarity(query_doc.getWordCounts()))

        self.similarities = similarities
        self.removeLastDocument()
        self.__assert__()

    # subscriptable and slice-able
    def __getitem__(self, n):
        return self.get(n)
    def getQuerySimilarity(self,idx:int):
        return self.similarities[idx]





        
        
        

    